plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs

```diff
@@ -0,0 +1,202 @@
+
+mkPTokenState :: VirtualFile -> PTokenState a
+mkPTokenState vf =
+  PTokenState
+    { rangeIdSetMap = mempty,
+      rope = Rope.fromText $ toText vf._file_text,
+      cursor = Char.Position 0 0,
+      columnsInUtf16 = 0,
+      currentRange = startRange,
+      currentRangeContext = NoSplit ("", startRange)
+    }
+
+addRangeIdSetMap :: (Monad m) => Range -> Identifier -> Tokenizer m ()
+addRangeIdSetMap r i = modify $ \s -> s {rangeIdSetMap = Map.insertWith (<>) r (S.singleton i) $ rangeIdSetMap s}
+
+-- lift a Tokenizer Maybe () to Tokenizer m (),
+-- if the Maybe is Nothing, do nothing, recover the state
+-- if the Maybe is Just (), do the action, and keep the state
+liftMaybeM :: (Monad m) => Tokenizer Maybe () -> Tokenizer m ()
+liftMaybeM p = do
+  st <- get
+  forM_ (execStateT p st) put
+
+hieAstSpanIdentifiers :: VirtualFile -> HieAST a -> RangeIdSetMap
+hieAstSpanIdentifiers vf ast = runIdentity $ runTokenizer (foldAst ast) (mkPTokenState vf)
+
+-- | foldAst
+-- visit every leaf node in the ast in depth first order
+foldAst :: (Monad m) => HieAST t -> Tokenizer m ()
+foldAst ast = if null (nodeChildren ast)
+  then liftMaybeM (visitLeafIds ast)
+  else mapM_ foldAst $ nodeChildren ast
+
+visitLeafIds :: HieAST t -> Tokenizer Maybe ()
+visitLeafIds leaf = liftMaybeM $ do
+  let span = nodeSpan leaf
+  (ran, token) <- focusTokenAt leaf
+  -- if `focusTokenAt` succeed, we can safely assume we have shift the cursor correctly
+  -- we do not need to recover the cursor state, even if the following computation failed
+  liftMaybeM $ do
+    -- only handle the leaf node with single column token
+    guard $ srcSpanStartLine span == srcSpanEndLine span
+    splitResult <-  lift $ splitRangeByText token ran
+    modify $ \s -> s {currentRange = ran, currentRangeContext = splitResult}
+    mapM_ combineNodeIds $ Map.filterWithKey (\k _ -> k == SourceInfo) $ getSourcedNodeInfo $ sourcedNodeInfo leaf
+  where
+    combineNodeIds :: (Monad m) => NodeInfo a -> Tokenizer m ()
+    combineNodeIds (NodeInfo _ _ bd) = mapM_ getIdentifier (M.keys bd)
+    getIdentifier :: (Monad m) => Identifier -> Tokenizer m ()
+    getIdentifier idt = liftMaybeM $ do
+      ran <- gets currentRange
+      case idt of
+        Left _moduleName -> addRangeIdSetMap ran idt
+        Right name -> do
+          ranSplit <- gets currentRangeContext
+          occStr <- lift $ case (occNameString . nameOccName) name of
+            -- the generated selector name with {-# LANGUAGE DuplicateRecordFields #-}
+            '$' : 's' : 'e' : 'l' : ':' : xs -> Just $ takeWhile (/= ':') xs
+            ['$']                            -> Just "$"
+            -- other generated names that should not be visible
+            '$' : _                          -> Nothing
+            ns                               -> Just ns
+          case ranSplit of
+            (NoSplit (tk, r)) -> do
+              guard $ T.unpack tk == occStr
```

Comment:
string comparisions can be expensive, using something like `Development.IDE.Types.Export.renderOcc = decodeUtf8 . bytesFS . occNameFS`, or even something like `encodeUtf8 text == bytesFS (occNameFS)` since encoding UTF8 with text 2.0+ is just a simple copy. Ideally we could avoid even the copy if we just compared the raw bytes in the `Text` and the `FastString` from `occNameFS` but this is not very easy to do.

Reply:
revised, see if current appoach would be better.

Reply:
I would try `encodeUtf8 text == bytesFS (occNameFS occ)` as that skips UTF-8 validation.

Reply:
We are comparing the extracted text(`Text`) from rope with the occStr(`Text`) that have stripped the generated prefix.
I am not sure it is going to be better if we `encodeUtf8` on both side and compares them.🤔